[CSSimplify] Delay binding metatype instance types if one side is type variable

xedin opened the pull request:
While matching two metatypes where one side is a class (which could have
supertypes) and another is a type variable use `subtype` constraint to
delay binding types together because there is a real possibility that
there exists a superclass associated with yet free type variable which
would be a better match when attempted.

Resolves: rdar://problem/44816848

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test

rudkx commented:
The metatype handling code here (prior to your change) seems pretty odd to me. Why do we default to a type equality check regardless of what the original type match kind was?

I wonder what breaks if we just make this immediately call back into `matchTypes` with the same `kind`, and the instance types?

jrose-apple commented on test/Constraints/rdar44816848.swift:
```diff
@@ -0,0 +1,13 @@
+}
+
+func bar(_ s: S!) {
+  s.foo(types: [A.self, B.self])
```
Nitpick: Since the bindings might not always be tried in the same order, having a second call with the array reversed seems like good future-proofing.

xedin replied:
I don't think element order here matters match because it used to binding array type to `A` and then equate to `B`, and reverse it would lead to the same result...

xedin replied:
I went ahead and added both reverse cases to `S!` and `S` just in case :)

xedin commented:
@rudkx I think the reason is trying to maintain subtype property? I wonder if what we need to to do here is to resolve this constraint (e.g. call to `matchTypes`) only if both types are known, otherwise just form unresolved, because currently e.g. in case of `ExistentialMetatype` only one side is checked...

xedin committed:
[CSSimplify] Delay binding metatype instance types if one side is type variable

While matching two metatypes where one side is a class (which could have
supertypes) and another is a type variable use `subtype` constraint to
delay binding types together because there is a real possibility that
there exists a superclass associated with yet free type variable which
would be a better match when attempted.

Resolves: rdar://problem/44816848

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -1962,21 +1962,32 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
+        // both underlying types are known because in
+        // case when inheritance involved we need to
+        // check subtype relation instead of equality.
+        if (instanceType1->isTypeVariableOrMember() ||
```
I think this is reasonable to check only for metatypes in this cause because existentials are going to form `subtype` anyway, WDYT?

xedin replied:
ok dah, there is an easier way to do this, I'm going to force push in a sec.

xedin commented:
@swift-ci please test

rudkx reviewed (approved):
Change LGTM. Might be good to follow-up at some point with an assert about the incoming `kind` so that if we ever violate any assumptions being made here it will be flagged.
